### ISSUE TITLE
API breaking changes proposal

### ISSUE THREAD

**Issue by frewsxcv** [frewsxcv/rust-threadpool#77]
* add `struct ThreadPoolBuilder`

--------

* add `ThreadPoolBuilder` `fn new()`
  * uses `num_cpu` crate to default num of threads to num of cpus

--------

* remove: `ThreadPool` `fn new(num_threads: usize) -> ThreadPool`
* add: `ThreadPool` `fn new() -> ThreadPool`
  * uses `num_cpu` to default num of threads to num of cpus
* add: `ThreadPoolBuilder` `fn num_threads(self, num: usize) -> ThreadPoolBuilder`
  * set the number of threads for the pool that the builder will construct

----------

* remove: `ThreadPool` `fn with_name(name: String, num_threads: usize) -> ThreadPool`
* add: `ThreadPoolBuilder` `fn name(self, name: String) -> ThreadPoolBuilder`

--------

* add: `ThreadPoolBuilder` `fn stack_size(self, size: usize) -> ThreadPoolBuilder`
  * https://doc.rust-lang.org/std/thread/struct.Builder.html#method.stack_size

---------

In summary: remove all constructors off of `ThreadPool`, except a new `new` constructor. If you want to fine-tune your threadpool, you'd use a `ThreadPoolBuilder`.

**Comment by frewsxcv** [comment]
It also might be good to unify language for setting the max number of threads the threadpool can start up. Maybe we should have something like:

* `ThreadPoolBuilder` `fn set_max_num_threads` (instead of ` fn num_threads`)
* `ThreadPool` `fn set_max_num_threads` (instead of `set_num_threads`)
* `ThreadPool` `fn max_num_threads` (instead of `max_count`)

or maybe we should just call these `num_threads` instead of `max_num_threads`

**Comment by dns2utf8** [comment]
I think a more exact/explicit wording is better for the builder.

**Comment by frewsxcv** [comment]
It crossed my mind that if the user needs `Clone` or `PartialEq`, the user could just wrap the `ThreadPool` in an `Arc` instead of relying on the internal one. Those traits seem like very niche use cases.

**Comment by frewsxcv** [comment]
Also, it might make sense to have `execute` and  `join` return a `Result` instead of sometimes panicking.

**Comment by dns2utf8** [comment]
The problem with putting the `ThreadPool` inside an `Arc` is, the `Sender<_>` is not `Send` and has checks against moving it from thread to thread. ([See this example](https://gist.github.com/dns2utf8/828d338bf3cb4214a8127786d89fe085))
Once the `TheadPool` is clonable, it feels natural that one can compare it with other instances.

I think `execute` can not fail because the `Receiver` will allways be there since the `ThreadPool` itself holds a reference to the shared data.

`join` may fail if for some reason a thread gets killed while also joining and thereby poisons the lock.

Did the two functions ever panic in one of your projects?
